control-plane-pki-operator/certificaterevocationcontroller/certificaterevocationcontroller.go

```diff
@@ -0,0 +1,683 @@
+package certificaterevocationcontroller
+
+import (
+	"context"
+	"crypto/sha256"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"math/big"
+	"reflect"
+	"sort"
+	"strings"
+	"time"
+
+	hypershiftv1beta1 "github.com/openshift/hypershift/api/hypershift/v1beta1"
+	hypershiftv1beta1applyconfigurations "github.com/openshift/hypershift/client/applyconfiguration/hypershift/v1beta1"
+	hypershiftclient "github.com/openshift/hypershift/client/clientset/clientset"
+	hypershiftinformers "github.com/openshift/hypershift/client/informers/externalversions"
+	"github.com/openshift/hypershift/control-plane-pki-operator/certificates"
+	"github.com/openshift/hypershift/control-plane-pki-operator/manifests"
+	"github.com/openshift/library-go/pkg/certs/cert-inspection/certgraphanalysis"
+	"github.com/openshift/library-go/pkg/certs/cert-inspection/certgraphapi"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/certrotation"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	corev1applyconfigurations "k8s.io/client-go/applyconfigurations/core/v1"
+	metav1applyconfigurations "k8s.io/client-go/applyconfigurations/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	certutil "k8s.io/client-go/util/cert"
+	"k8s.io/klog/v2"
+)
+
+type CertificateRevocationController struct {
+	kubeClient       kubernetes.Interface
+	hypershiftClient hypershiftclient.Interface
+
+	fieldManager string
+	getCRR       func(namespace, name string) (*hypershiftv1beta1.CertificateRevocationRequest, error)
```

Comment:
openshift and kube just keep the lister and unit test by creating an indexer.  This is for unit testing without a lister?  openshift moved away from that approach shortly after v3.0, why bring it back?

Reply:
My experience with fake clients is almost uniquely negative and frustrating.